katago_server/distributed_efforts/services/rating_network_pairer.py

```diff
@@ -68,10 +75,23 @@ def _choose_opponent(self, reference_network):
             .filter(run=self.current_run, log_gamma__lte=log_gamma_upper_bound, log_gamma__gte=log_gamma_lower_bound,)
             .all()
         )
-        if len(nearby_networks) < 5:
-            nearby_networks_a = Network.objects.filter(run=self.current_run, pk__lt=reference_network.pk).all()[:20]
-            nearby_networks_b = Network.objects.filter(run=self.current_run, pk__gt=reference_network.pk).all()[:20]
-            return np.random.choice(nearby_networks_a + nearby_networks_b)
+        if len(nearby_networks) < 4:
+            nearby_weaker_networks = (
+                Network.objects.exclude(pk=reference_network.pk)
+                .filter(run=self.current_run, log_gamma__lte=ref_net_log_gamma)
+                .order_by("-log_gamma")
+                .all()[:2]
```

Comment:
I thin the right logic to `find the X closest network in elo of network Y` should be:
- find X (and not X /2) networks after Y, ranked by elo excluding Y
- find X (and not X /2) networks before Y, ranked by elo excluding Y
- compute elo diff
- order by abs(elo_diff)
- select first X

That will works if all the network are above elo Y.

In this case it means that it is 4 and not 2.
And Since it start being a complex logic, I would extract it into a querySet or a manager, like current_run function for instance.

Reply:
I wasn't attempting to find precisely the X closest, just the nearest two on each side I think is fine anyways. I think this is just a corner case where doing anything reasonable is okay, it won't be very important. Except for maybe a random network or the first network after random, it will never be the case networks are more than 1200 Elo apart.

Reply:
Fine for me.